### PR TITLE
[Stage][Fix] Add additional conditions when checking types of output from the model

### PR DESCRIPTION
**Pull request by SeanNaren**
Fixes an upstream issue in Lightning that I found when testing the latest DeepSpeed version on our CI.

From the forward pass in Lightning, we offer a few different potential output types. Due to the additional type checking of the output, some of these failed (`None` and `Dict`). This adds additional checks to ensure we handle those cases within the hook.

If possible I'd like to add a test (even though Lightning CI caught this), any guidance on where would be the right place?

### CODE REVIEW

File: tests/unit/test_zero_context.py [hunk 1]
```diff
@@ -243,3 +243,42 @@ def test_ext_param_returnobj():
...
+        self.dim = dim
+        self.linear1 = torch.nn.Linear(dim, dim)
+
+    def forward(self, input):
+        act1 = self.linear1(input)
+        if self.output_type is None:
+            return None
+        if self.output_type is dict:
+            return {'loss': act1.sum()}
+        return act1.sum()
```
Comment (tjruwase) [inline tests/unit/test_zero_context.py:261]: It seems this code path will not be executed since the output types are restricted to dict and None?

Comment (SeanNaren) [reply tests/unit/test_zero_context.py:261]: Yep totally right, I can remove this (wanted to keep it general)

Comment (tjruwase) [reply tests/unit/test_zero_context.py:261]: It is good to keep it general, then perhaps create a test case with a different output type?

Comment (SeanNaren) [reply tests/unit/test_zero_context.py:261]: Done @tjruwase let me know what you think :)

**Review by tjruwase — APPROVED** [review]

**Pull request merged**